### ISSUE TITLE
Investigate and Optimize Egress Costs on EU1 Satellite due to Excessive ListObject Requests

### ISSUE THREAD

**Issue by NiaStorj** [storj/edge#503]

**Comment by onionjake** [comment]
I've rate limited the accounts causing the excessive list requests, @mniewrzal suggests that gateway-mt might not need to return nearly as much metadata as it does, which might reduce egress costs.

> one issue with ListObjects is that edge is listing with included full metadata which because of some compatibility  problems from the past

**Comment by amwolff** [comment]
ListObjects lists including full metadata (system & custom) because to be compatible with S3, we need to return size and modtime (system) as well as ETag (saved in custom on upload) /cc @mniewrzal

**Comment by amwolff** [comment]
A little investigation; here's what I discovered:

This is the number of exhaustive listings triggered per second:

![image](https://github.com/user-attachments/assets/79a6e981-a525-4a62-9b06-8a0bcc92f12e)

Link: https://thanos.storj.rodeo/goto/VtVK30eSR?orgId=1

What it means is that the extreme amount of traffic comes from the inefficient listing code path that the customer is triggering (using non-standard prefix/delimiter).

A quick solution is to disable exhaustive listing or reduce the size of the prefix it can be triggered on (currently, exhaustive listing works on prefixes which have as much as 100k items following them).

Disabling it completely will of course break compatibility. A more elegant solution is to:

1. Make arbitrary prefixes/delimiters work on the satellite when path encryption is disabled
2. Adjust the gateway
3. Teach customers to use that if they want more powerful listing controls
4. Remove exhaustive listing from the gateway

Unfortunately, this is a much bigger undertaking than just limiting the impact by adjusting the dial on the gateway. @onionjake do you want us to take action, or did you just want to know where this comes from?

**Comment by mniewrzal** [comment]
> ListObjects lists including full metadata (system & custom) because to be compatible with S3, we need to return size and modtime (system) as well as ETag (saved in custom on upload) /cc @mniewrzal

Ah I completely forgot about ETag. That might complicate things a lot. I need to think for a moment.

**Comment by amwolff** [comment]
Yes, exactly, if they use an arbitrary prefix/delimiter, then one ListObjects(V2) call can map to thousands of requests against their rate limit.
